src/timeaxis/timegrid.jl

```diff
@@ -287,6 +299,45 @@ Base.resize!(tg::TimeGrid{T,P,:infinite}, n::Int) where {T,P} = tg
 
 Base.diff(tg::TimeGrid{T,P,:finite}) where {T,P} = fill(tg.p, tg.n - 1)
 
+###############################################################################
+#  Intersect
+###############################################################################
+function Base.intersect(tg1::TimeGrid, tgs...)
+    for tg in tgs
+        tg1 = intersect(tg1, tg)
+    end
+    return tg1
+end
+# TODO 
+# Consider checking if they are equal
+function intersect(tg_inf::TimeGrid{T,P,:infinite}, 
+                   tg_fin::TimeGrid{T,P,:finite})::TimeGrid{T,P,:finite} where {T,P}
+    @assert tg_fin.p % tg_inf.p
+    if tg_fin[end] < tg_inf.o
+        return TimeGrid(tg_fin.o, tg_fin.p, 0)
+    end
+    new_o = max(tg_fin.o, tg_inf.o)
+    new_n = tg_inf.o > tg_fin.o ? findfirst(isequal(tg_fin[end]), tg_inf) : 
+                                  findfirst(isequal(tg_fin[end]), tg_fin)
+    return TimeGrid(new_o, tg_fin.p, new_n)
+end
+intersect(tg_fin::TimeGrid{T,P,:finite}, tg_inf::TimeGrid{T,P,:infinite}) where {T,P} = intersect(tg_inf, tg_fin)
+function intersect(tg_inf1::TimeGrid{T,P,:infinite}, tg_inf2::TimeGrid{T,P,:infinite}) where {T,P}
+    @assert tg_inf1.p == tg_inf2.p
+    new_o = max(tg_inf1.o, tg_inf2.o)
+    return TimeGrid(new_o, tg_inf1.p)
+end
+function intersect(tg_fin1::TimeGrid{T,P,:finite}, tg_fin2::TimeGrid{T,P,:finite}) where {T,P}
```

Comment:
I found a case that might yield empty TimeGrid:
```julia
julia> tg = TimeGrid(DateTime(2021, 1, 1, 0), Hour(1), 24)
24-element TimeGrid{DateTime, Hour, :finite}:
 2021-01-01T00:00:00
  ⋮
 2021-01-01T23:00:00
 / 1 hour

julia> tg′ = TimeGrid(DateTime(2021, 1, 1, 0, 30), Hour(1), 24)
24-element TimeGrid{DateTime, Hour, :finite}:
 2021-01-01T00:30:00
  ⋮
 2021-01-01T23:30:00
 / 1 hour

julia> intersect(collect(tg), collect(tg′))
DateTime[]

```

Reply:
In fact I was wrong on slack there is a way to infinite with infinite intersection being empty

Reply:
I rewrote some parts and added more tests. 

The current behavior of intersecting tg with different periods is method error. What do you think about that?

Reply:
> The current behavior of intersecting tg with different periods is method error. What do you think about that?

That might need `IrregularTimeGrid` implemented.
I think we can revisit it later, putting a `TODO` here is fine for me.

I still don't make a perfect algo come out. Here is my two cents:
- For the case `tg.p <: Period` (or maybe narrow down to `FixedPeriod` ?), treat all the period as `Int` and we can get a `TimeGrid`.
- For the case `tg.p <: Float64` (and `Irrational` included. Although I know `π` is marked a `Irrational` type, all calculation on it are approximations.), just fallback to the `intersect(::Vector, ::Vector)` calculation and output `IrregularTimeGrid`.